File: 210-course-schedule-ii/course-schedule-ii.java
class Solution {
    public int[] findOrder(int n, int[][] prerequisites) {
        int[] ans = new int[n] ;
        int[] in = new int[n] ;
        Queue<Integer> q = new LinkedList<>() ;
        Map<Integer,List<Integer>> map = new HashMap<>() ;
        for(int i = 0 ; i<n ; i++)map.put(i,new ArrayList<>()) ;
        for(int i = 0 ; i<prerequisites.length ; i++){
            map.get(prerequisites[i][1]).add(prerequisites[i][0]) ;
            in[prerequisites[i][0]]++ ;
        }
        int idx = 0 ;
        for(int i = 0 ; i<n ; i++){
            if(in[i]==0){
                q.add(i) ;
                ans[idx++] = i; 
            }
        }
        while(!q.isEmpty()){
            int cur = q.poll() ;
            for(int i : map.get(cur)){
                in[i]-- ;
                if(in[i]==0){
                    q.add(i) ;
                    ans[idx++] = i ;
                }
            }
        }
        return idx!=n ? new int[]{} : ans ;
    }
}

File: 210-course-schedule-ii/course-schedule-ii.cpp
class Solution {
public:
    vector<int> findOrder(int n, vector<vector<int>>& prerequisites) {
        vector<int> ans(n);
        vector<int> in(n, 0);
        queue<int> q;
        unordered_map<int, vector<int>> map;

        for (int i = 0; i < n; i++) {
            map[i] = vector<int>();
        }

        for (int i = 0; i < prerequisites.size(); i++) {
            map[prerequisites[i][1]].push_back(prerequisites[i][0]);
            in[prerequisites[i][0]]++;
        }

        int idx = 0;

        for (int i = 0; i < n; i++) {
            if (in[i] == 0) {
                q.push(i);
                ans[idx++] = i;
            }
        }

        while (!q.empty()) {
            int cur = q.front();
            q.pop();

            for (int i : map[cur]) {
                in[i]--;
                if (in[i] == 0) {
                    q.push(i);
                    ans[idx++] = i;
                }
            }
        }

        if (idx != n) return {};

        return ans;
    }
};